plasmapy/diagnostics/thomson.py

```diff
@@ -318,4 +393,254 @@ def spectral_density(
     # Recast as real: imaginary part is already zero
     Skw = np.real(np.sum(econtr, axis=0) + np.sum(icontr, axis=0))
 
+    # Apply an insturment function if one is provided
+    if inst_fcn is not None and callable(inst_fcn):
+        # Create an array of wavelengths of the same size as wavelengths
+        # but centered on zero
+        wspan = (np.max(wavelengths) - np.min(wavelengths)) / 2
+        eval_w = np.linspace(-wspan, wspan, num=wavelengths.size)
+        # Evaluate the insturment function
+        inst_fcn_arr = inst_fcn(eval_w)
+        # Ensure inst_fcn is normalized
+        inst_fcn_arr *= 1 / np.sum(inst_fcn_arr)
+        # Convolve the insturment function with the spectral density
+        # linear rather than circular convolution is important here!
+
+        # TODO: Is numpy, scipy.signal, or astropy the best convolution
+        # function here??
+        Skw = np.convolve(Skw.value, inst_fcn_arr.value, mode="same") * Skw.unit
+
     return np.mean(alpha), Skw
+
+
+# ***************************************************************************
+# These functions are necessary to interface scalar Parameter objects with
+# the array inputs of spectral_density
+# ***************************************************************************
+
+
+def _count_populations_in_params(keys, prefix):
+    """
+    Counts the number of entries matching the pattern prefix_i in a
+    list of keys
+    """
+
+    i = 0
+    while True:
+        if prefix + f"_{i}" in keys:
+            i += 1
+        else:
+            break
+    return i
```

Comment:
This is a bit brutish.  I'm curious if this could be done more cleanly with functionality from python's [`re`](https://docs.python.org/3/library/re.html) package.

Reply:
Good idea - turned out to be a one-liner. 
```python
return len(re.findall(prefix, ",".join(keys)))
```
I tend to avoid regex because I don't _actually_ understand it half the time.

Reply:
I think that's why most people avoid regex.  I always think to myself, "oh man, I have to relearn this now!" 😂